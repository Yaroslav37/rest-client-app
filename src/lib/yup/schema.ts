import * as yup from 'yup';

const hasNumber = /\p{N}/u;
const hasUpperCase = /\p{Lu}/u;
const hasLowerCase = /\p{Ll}/u;
const hasSpecialChar = /[^\p{L}\p{N}]/u;
const minLength = 8;

export const validationSchema = yup.object().shape({
  email: yup.string().email('Invalid email').required('Email is required'),
  password: yup
    .string()
    .matches(hasNumber, 'Password must contain at least 1 number')
    .matches(hasUpperCase, 'Password must contain at least 1 uppercase letter')
    .matches(hasLowerCase, 'Password must contain at least 1 lowercase letter')
    .matches(hasSpecialChar, 'Password must contain at least 1 special character')
    .min(minLength, `Password must be at least ${minLength} characters`)
    .required('Password is required'),
  passwordConfirm: yup
    .string()
    .oneOf([yup.ref('password')], 'Passwords must match')
    .required('Confirm password is required'),
});
