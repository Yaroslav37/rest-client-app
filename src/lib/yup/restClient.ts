import * as yup from 'yup';

import { HTTP_METHODS } from '@/shared/constants';

export const restClientSchema = yup.object({
  method: yup.string().required('Method is required').oneOf(HTTP_METHODS, 'Invalid HTTP method'),
  url: yup
    .string()
    .required('URL is required')
    .url('Invalid URL format')
    .test(
      'no-variables',
      'Variables must be replaced before sending',
      (value) => !/\{\{.+?\}\}/.test(value),
    ),
  headers: yup.array().of(
    yup.object({
      key: yup.string().required('Header key is required'),
      value: yup.string().required('Header value is required'),
      id: yup.number().required('Header id is required'),
    }),
  ),
  body: yup.string().test('valid-json', 'Invalid JSON format', (value) => {
    if (!value) return true;
    try {
      JSON.parse(value);
      return true;
    } catch {
      return false;
    }
  }),
});

export type RestClientFormValues = yup.InferType<typeof restClientSchema>;
