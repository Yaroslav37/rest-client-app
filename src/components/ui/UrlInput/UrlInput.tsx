'use client';

import { Control, Controller } from 'react-hook-form';

import { RestClientFormValues } from '@/lib/yup/restClient';

interface Props {
  control: Control<RestClientFormValues>;
}

export const UrlInput = ({ control }: Props) => {
  return (
    <Controller
      name="url"
      control={control}
      render={({ field }) => (
        <input
          {...field}
          type="text"
          className="text-white"
          placeholder="Enter endpoint URL"
          onChange={(e) => {
            field.onChange(e.target.value);
          }}
        />
      )}
    />
  );
};
