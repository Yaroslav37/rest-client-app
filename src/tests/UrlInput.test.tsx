import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { useForm } from 'react-hook-form';
import { describe, expect, it, vi } from 'vitest';

import { UrlInput } from '@/components/ui/UrlInput/UrlInput';
import { RestClientFormValues } from '@/lib/yup/restClient';

vi.mock('next-intl', () => ({
  useTranslations: () => (key: string) => {
    const translations: Record<string, string> = {
      'RestClient.enter-url': 'Enter URL',
    };
    return translations[key] || key;
  },
}));

describe('UrlInput Component', () => {
  const Wrapper = () => {
    const { control } = useForm<RestClientFormValues>({
      defaultValues: { url: '' },
    });
    return <UrlInput control={control} />;
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  it('renders correctly with label and input', () => {
    render(<Wrapper />);

    expect(screen.getByText('enter-url')).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toBeInTheDocument();
    expect(screen.getByRole('textbox')).toHaveAttribute(
      'placeholder',
      'https://jsonplaceholder.typicode.com/posts',
    );
  });

  it('applies correct styling classes', () => {
    render(<Wrapper />);

    const input = screen.getByRole('textbox');
    expect(input).toHaveClass('text-white');
    expect(input).toHaveClass('border-b');
    expect(input).toHaveClass('border-green');
  });

  it('updates value when typing', async () => {
    render(<Wrapper />);
    const input = screen.getByRole('textbox');
    await userEvent.type(input, 'https://example.com/api');
    expect(input).toHaveValue('https://example.com/api');
  });

  it('matches snapshot', () => {
    const { container } = render(<Wrapper />);
    expect(container).toMatchSnapshot();
  });
});
