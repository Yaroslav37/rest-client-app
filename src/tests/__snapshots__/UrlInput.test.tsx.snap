// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UrlInput Component > matches snapshot 1`] = `
<div>
  <div
    class="flex gap-2 flex-col justify-between font-montserrat max-w-[375px] sm:w-full"
  >
    <span
      class="text-light-green self-start sm:self-end"
    >
      enter-url
    </span>
    <input
      class="text-white p-2 border-b border-green focus:outline-0 focus:border-white"
      name="url"
      placeholder="https://jsonplaceholder.typicode.com/posts"
      type="text"
      value=""
    />
  </div>
</div>
`;
